tests/functional/test_freeze.py

```diff
@@ -88,11 +88,29 @@ def test_basic_freeze(script: PipTestEnvironment) -> None:
 
 
 def test_freeze_with_pip(script: PipTestEnvironment) -> None:
-    """Test pip shows itself"""
+    """Test that pip shows itself only when --all is used"""
+    result = script.pip("freeze")
+    assert "pip==" not in result.stdout
     result = script.pip("freeze", "--all")
     assert "pip==" in result.stdout
 
 
+def test_freeze_with_setuptools(script: PipTestEnvironment) -> None:
+    """
+    Test that pip shows setuptools only when --all is used
+    or Python version is >=3.12
+    """
+
+    result = script.pip("freeze")
+    if sys.version_info >= (3, 12):
```

Comment:
Can we do this test by mocking the value of `sys.version_info`, rather than having the test do different things depending on what Python version it's run under? That may be overkill, and may be difficult to achieve because the test runs pip in a subprocess, so I'm not suggesting that this is essential, but it's worth thinking about (especially as CI currently won't check this change at all).

Reply:
Seems like overkill, frankly. 3.12 will be added to CI sooner or later anyway.

Reply:
+1 to mocking, doing conditional here makes running tests locally unnecessarily difficult.

Reply:
I gave it more thought, and mocking `sys.version_info` really does not seem like a good idea. Imagine that pip (or one of its dependencies) does something like this in the code:

```python
if sys.python_version >= (3, 12):
    # use feature that is only available in 3.12+
```

Then if you set `sys.python_version` to 3.12.x while running on, say, 3.8.x, then this code will break.

Reply:
Or two tests with different skip_if decorators?

Reply:
Instead of mocking `sys.version_info` directly, I’d add a function that toggles freeze output, and mock that.

Reply:
> Instead of mocking `sys.version_info` directly, I’d add a function that toggles freeze output, and mock that.

Okay, I did it.